### ISSUE TITLE
Doubts regarding inductive records with eta

### ISSUE THREAD

**Issue by anuyts** [agda/agda#5842]
I am on Agda version 2.6.2-246a229.

I just found out that Agda has inductive records with eta-equality.
So I thought: I'll implement natural numbers as an inductive record, define the identity function recursively, and prove that it is equal to `λ x -> x` by writing `refl`. Something is bound to go wrong, and then I'll write a bug report.

Things went wrong early. In the following code:
```agda
open import Data.Bool
open import Data.Unit
open import Data.Empty

record Nat' : Set where
  inductive
  eta-equality
  constructor nat'
  field
    isZero : Bool
    pred : (if isZero then ⊥ else ⊤) → Nat'

open Nat'

idNat1 : Nat' → Nat'
isZero (idNat1 n) = isZero n
pred (idNat1 n) x = idNat1 (pred n {!x!})

idNat2 : Nat' → Nat'
idNat2 (nat' false f) = nat' false λ x → idNat2 (f x)
idNat2 (nat' true f) = nat' true λ()
```
both `idNat1` and `idNat2` fail the termination checker (see #5823 I guess). What's the point of an inductive type if you cannot recurse over it?
Moreover, in `idNat1`, I cannot *give* `x` because the prior copattern does not reduce in the goal of the second one (see #2068 although that issue supposedly has been fixed).

So I thought, I'll give `idNat1` another try by removing the option to have a zero. My experiment should still be possible:
```agda
record NoNat : Set where
  inductive
  eta-equality
  constructor suc
  field
    pred : NoNat

open NoNat

idNoNat : NoNat→ {!NoNat!}
idNoNat = {!!}
```
This still typechecks, but if you take the last `NoNat` out of its hole, i.e. `idNoNat : NoNat→ NoNat`, then it loops.

Anyway, if it works as expected, then I would continue to write
```agda
record NoNat : Set where
  inductive
  eta-equality
  constructor suc
  field
    pred : NoNat

open NoNat

{-# TERMINATING #-} -- it is!
idNoNat : NoNat → NoNat
pred (idNoNat n) = idNoNat (pred n)

open import Relation.Binary.PropositionalEquality

idNoNatEq : (n : NoNat) → idNoNat n ≡ n
idNoNatEq (suc n) = refl
```
which should typecheck perfectly fine, although at the same time it is impossible.

**Comment by anuyts** [comment]
Perhaps it doesn't make sense to corecurse on an inductive type, but you could have
```agda
{-# TERMINATING #-} -- it is!
idNoNat : NoNat → NoNat
idNoNat (suc n) = suc (idNoNat n)
```

**Comment by andreasabel** [comment]
Thanks @anuyts, for the report.  ~~Let me answer once I have finalized my fix for #5823.~~

**Comment by andreasabel** [comment]
@anuyts wrote:
> ```agda
> record Nat' : Set where
>   inductive
>   eta-equality
>   constructor nat'
>   field
>     isZero : Bool
>     pred : (if isZero then ⊥ else ⊤) → Nat'
> 
> idNat1 : Nat' → Nat'
> isZero (idNat1 n) = isZero n
> pred (idNat1 n) x = idNat1 (pred n {!x!})
> 
> idNat2 : Nat' → Nat'
> idNat2 (nat' false f) = nat' false λ x → idNat2 (f x)
> idNat2 (nat' true f) = nat' true λ()
> ```
> 
> both `idNat1` and `idNat2` fail the termination checker (see #5823 I guess). 

They should.  If you were to except `idNat1` on the basis of guardedness (reduced observation depth), then you could construct an inhabitant of `NoNat` as well, but `NoNat` is an empty type.

Also, `idNat2` is rewritten to use of projections on the rhs rather than matching on the lhs (because you wanted an eta record).  After this transformation, it is clearly looping.  Ok, turns out that this transformation is actually not possible because you are matching on a `Bool`.  But the termination checker does not cater for such subtleties.  It simply does not regard the subterms of an eta-constructor as smaller than the whole thing, so there is no basis for justifying termination.

> What's the point of an inductive type if you cannot recurse over it? 

`inductive` is here not really _inductive_, but the antagonist of `coinductive` (which changes the reduction semantics, I think, but I cannot remember the exact behavior now).

The main application seems to be "terminating records" stuff like (taken from the standard library):
```agda
  record Record {s} (Sig : Signature s) : Set s where
    eta-equality
    inductive
    constructor rec
    field fun : Record-fun Sig

  Record-fun : ∀ {s} → Signature s → Set s
  Record-fun ∅             = ⊤
  Record-fun (Sig , ℓ ∶ A) =          Σ (Record Sig) A
  Record-fun (Sig , ℓ ≔ a) = Manifest-Σ (Record Sig) a
```
You need eta to work with these:
```agda
{-# OPTIONS --with-K #-}

open import Data.Product
open import Data.String
open import Function.Base using (flip)
open import Level
open import Relation.Binary

import Data.Record as Record

-- Let us use strings as labels.

open Record String _≟_

-- Partial equivalence relations.

PER : Signature _
PER = ∅ , "S"     ∶ (λ _ → Set)
        , "R"     ∶ (λ r → r · "S" → r · "S" → Set)
        , "sym"   ∶ (λ r → Lift _ (Symmetric (r · "R")))
        , "trans" ∶ (λ r → Lift _ (Transitive (r · "R")))

-- Given a PER the converse relation is also a PER.

converse : (P : Record PER) →
           Record (PER With "S" ≔ (λ _ → P · "S")
                       With "R" ≔ (λ _ → flip (P · "R")))
converse P =
  rec (rec (_ ,
    lift λ {_} → lower (P · "sym")) ,
    lift λ {_} yRx zRy → lower (P · "trans") zRy yRx)
```
Without eta, the `_` is not solved.

**Comment by anuyts** [comment]
So is it correct to say that "inductive records" are intended to define types *by recursion on something* (e.g. `Record` is defined by induction on a `Signature`) and to be used by induction on that thing (by induction on the `Signature`)?

The strange thing is that it is still possible to define inductive *types* (such as the natural numbers) as inductive *records* with eta, even though the expected behaviour for such a situation is impossible and therefore also disabled. I'm not sure if a warning is necessary (along the lines of "You will not be able to recurse on this type" when you're not really recursing over the indices) but at least the documentation should clarify the intended use case.

Does the fix of #5823 fix the looping I got above for `NoNat`?

**Comment by anuyts** [comment]
Actually I'm not convinced that `Data.Record` is a relevant use case at all. Since `Signature` is an inductive type, signatures are always of finite length, so records would have finite length even if defined coinductive. To my understanding, inductive records have strictly less functionality than coinductive types, so there seems to be no reason not to define `Record` as coinductive.

**Comment by andreasabel** [comment]
> This still typechecks, but if you take the last `NoNat` out of its hole, i.e. `idNoNat : NoNat→ NoNat`, then it loops.
> ...
> Does the fix of https://github.com/agda/agda/issues/5823 fix the looping I got above for `NoNat`?

Indeed, this no longer loops (the fix of #5823 has now merged into `master`).

> The strange thing is that it is still possible to define inductive _types_ (such as the natural numbers) as inductive _records_ with eta, even though the expected behaviour for such a situation is impossible and therefore also disabled. I'm not sure if a warning is necessary (along the lines of "You will not be able to recurse on this type" when you're not really recursing over the indices) but at least the documentation should clarify the intended use case.

With the new termination check for records, it might be worthwhile to review and document the flavors of `record` types Agda supports.  These have just seen refinement after refinement, now we have these dimensions:
1. `inductive` or `coinductive`
2. recursive? (inferred by positivity checker)
3. terminating? (inferred by termination checker)
4. `eta-equality` or `no-eta-equality`
5. allow `pattern` matching?
6. allow `copattern` matching?

These attributes are not freely given of course, there are interconnections we should map out and document clearly.

- non-recursive => terminating
- ...
